### PR TITLE
Increase protocol version to 8 and add negotiation

### LINKED ISSUE #16459: Incompatible changes were made to the server order format without increasing the protocol version

### PR DESCRIPTION
**Pull request by jrb0001**
Fixes #16459.

### CODE REVIEW

**Comment by jrb0001** [comment]
Interop tests showed a small bug in the version check which has been fixed. It should be ready to be merged now.

File: OpenRA.Game/Server/Connection.cs [hunk 1]
```diff
@@ -90,6 +90,25 @@ public void ReadData(Server server)
 					var bytes = PopBytes(ExpectLength);
 					switch (State)
 					{
+						case ReceiveState.Handshake:
+							{
+								byte min = bytes[0];
```
Comment (pchote) [inline OpenRA.Game/Server/Connection.cs]: Use `var` instead of `byte` for these two lines.

File: OpenRA.Game/Server/Connection.cs [hunk 2]
```diff
@@ -90,6 +90,25 @@ public void ReadData(Server server)
...
+									Log.Write("server", "Dropping client {0} for unsupported version: Server={1}, Client Min={2}, Client Max={3}",
+											PlayerIndex, ProtocolVersion.Version, min, max);
+									server.DropClient(this);
+								}
+								else
+								{
+									ExpectLength = 8;
+									State = ReceiveState.Header;
+								}
+							} break;
```
Comment (pchote) [inline OpenRA.Game/Server/Connection.cs]: Please put this `break` inside the braces, and would you mind fixing the other two cases in this switch too?

This then lets me drop a couple more style fixes from my pending StyleCop PR.

**Review by pchote — COMMENTED** [review]
The logic implemented here looks fine, just a few style nits:

**Comment by reaperrr** [comment]
Also needs a rebase.

File: OpenRA.Game/Server/Connection.cs (hunk 2 above)
Comment (jrb0001) [reply OpenRA.Game/Server/Connection.cs]: I dropped the braces completely, it looks better that way and also matches how switch is used in other places.

**Comment by jrb0001** [comment]
Fixups and rebase done.

**Comment by pchote** [comment]
Travis failed:
```
Checking for code style violations in OpenRA.Game...
OpenRA.Game/Server/Connection.cs:L108: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.
OpenRA.Game/Server/Connection.cs:L121: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.
OpenRA.Game/Server/Server.cs:L281: [SA1515] A single-line comment must be preceded by a blank line or another single-line comment, or must be the first item in its scope. To ignore this error when commenting out a line of code, begin the comment with '////' rather than '//'.
Makefile:178: recipe for target 'check' failed
make: *** [check] Error 1
```

**Comment by jrb0001** [comment]
Fixed.

**Review by pchote — DISMISSED** [review]

**Comment by pchote** [comment]
Discussion in IRC raised an important point here: the protocol bump is going to break our mod-switching functionality between playtest and release versions (and different mods).

If we kept the protocol datatype as an int we could change `NetworkConnectionReceive` to read the minversion, and then only read the maxversion if minversion > 7. We could then update the order unpacking on both server and client to be able to deserialize protocol 7 without much hassle. We would lose important order information if two protocol 8 clients tried to play on a protocol 7 server, but the mod-version checks would prevent this - all we care about is that a version 8 client can decode the initial handshake from a version 7 server.

Adding compatibility for a version 7 client connecting to a version 8 server is trickier but perhaps not impossible. The client will want to interpret the max version int as the length for the first packet - so the server could wait for a second or two, and if the client doesn't respond with its own protocol versions the server could assume that it is 7, and send 11 dummy bytes that the client would interpret as a packet (and hopefully ignore) before sending the handshake.

I'm not immediately whether it is possible to build a dummy packet in this way which the client would silently ignore.

 then send some dummy data to turn it into an order packet

**Comment by pchote** [comment]
#16531 should now cover this without breaking the mod switcher, so closing as superseded.

**Pull request closed without merging**